File: api/services/analysis_validator.py
from typing import Dict, Any
from config.config import logger


def validate_analysis(analysis: Dict[str, Any]) -> None:
    """Проверяет наличие ключевых полей в разделах анализа.

    Для каждого отсутствующего поля выводится предупреждение через logger.
    """
    if not isinstance(analysis, dict):
        logger.warning("validate_analysis: входные данные должны быть словарем")
        return

    sections = {
        'primary_analysis': ['global_trend', 'local_trend', 'patterns', 'anomalies'],
        'feedback': ['issues', 'missed_data', 'model_configuration', 'note', 'time_period'],
        'confidence_in_trading_decisions': ['confidence', 'reason'],
        'indicator_correlations': ['macd_rsi_correlation', 'atr_volatility_correlation', 'explanation'],
        'volatility_by_intervals': ['morning_volatility', 'evening_volatility', 'comparison'],
        'market_cycles_identification': [],
        'extended_ichimoku_analysis': ['conversion_base_line_cross', 'price_vs_cloud'],
        'support_resistance_levels': ['supports', 'resistances'],
        'trend_lines': ['lines'],
        'unfinished_zones': ['type', 'level', 'date'],
        'imbalances': ['type', 'start_point', 'end_point', 'price_range'],
        'fibonacci_analysis': ['based_on_local_trend', 'based_on_global_trend'],
        'elliott_wave_analysis': ['current_wave', 'wave_count', 'forecast'],
        'structural_edge': ['type', 'date', 'price'],
        'candlestick_patterns': ['type'],
        'divergence_analysis': ['indicator', 'type', 'date'],
        'fair_value_gaps': ['date', 'price_range'],
        'gap_analysis': ['gaps', 'comment'],
        'psychological_levels': ['levels'],
        'anomalous_candles': ['date', 'type', 'price'],
        'price_prediction': ['forecast', 'virtual_candles'],
        'recommendations': ['trading_strategies'],
        'volume_analysis': ['volume_trends', 'significant_volume_changes'],
    }

    for key, fields in sections.items():
        value = analysis.get(key)
        if not value:
            continue

        if isinstance(value, list):
            for i, item in enumerate(value):
                if not isinstance(item, dict):
                    logger.warning(f"Раздел {key}: запись {i} имеет некорректный формат")
                    continue
                for field in fields:
                    if field not in item:
                        logger.warning(f"Раздел {key}: запись {i} отсутствует поле {field}")
        elif isinstance(value, dict):
            for field in fields:
                if field not in value:
                    logger.warning(f"Раздел {key}: отсутствует поле {field}")
            if key == 'gap_analysis' and isinstance(value.get('gaps'), list):
                for i, gap in enumerate(value['gaps']):
                    if not isinstance(gap, dict):
                        logger.warning(f"gap_analysis: запись {i} имеет некорректный формат")
                        continue
                    if 'start_point' not in gap or 'end_point' not in gap:
                        logger.warning(f"gap_analysis: запись {i} не содержит start_point/end_point")
        else:
            logger.warning(f"Раздел {key} имеет неожиданный тип {type(value).__name__}")

File: frontend/src/analysisValidator.js
// src/data/analysisValidator.js

/**
 * Проверяет структуру JSON-ответа от модели.
 * Для каждого раздела из списка sectionRequirements убеждается,
 * что присутствуют все необходимые поля.
 * В случае отсутствия — выводит console.warn, но не бросает ошибку.
 */

const sectionRequirements = {
  // Primary analysis
  primary_analysis: ['global_trend', 'local_trend', 'patterns', 'anomalies'],

  // Confidence
  confidence_in_trading_decisions: ['level', 'reason'],

  // Model Analysis layers
  support_resistance_levels: ['date', 'level', 'explanation'],
  trend_lines: ['start_point', 'end_point', 'explanation'],
  fibonacci_analysis: ['start_point', 'end_point', 'levels', 'explanation'],
  unfinished_zones: ['start_point', 'end_point', 'explanation'],
  imbalances: ['start_point', 'end_point', 'explanation'],
  fair_value_gaps: ['start_point', 'end_point', 'explanation'],
  gap_analysis: ['gaps'],  // специальная обработка ниже
  structural_edge: ['date', 'price', 'explanation'],
  candlestick_patterns: ['date', 'type', 'explanation'],
  divergence_analysis: ['indicator', 'type', 'date', 'explanation'],

  // Other analysis sections
  pivot_points: ['daily', 'weekly', 'monthly'],
  indicators_analysis: [],        // произвольная структура, просто проверяем наличие
  volume_analysis: ['volume_trends', 'significant_volume_changes'],
  indicator_correlations: ['correlations', 'explanation'],
  psychological_levels: ['level', 'date', 'type', 'explanation'],
  extended_ichimoku_analysis: ['details'],
  volatility_by_intervals: ['intervals'],
  anomalous_candles: ['date', 'price', 'explanation'],

  // Forecast & Recommendations
  price_prediction: ['virtual_candles'],
  recommendations: ['recommendations'],

  // Risk & Feedback
  risk_management: ['rules'],
  feedback: ['note', 'model_configuration', 'missed_data', 'issues_encountered', 'suggestions']
};

/**
 * Основная функция валидации.
 * @param {object} analysis — распарсенный JSON-ответ модели.
 */
export function validateAnalysis(analysis) {
  Object.entries(sectionRequirements).forEach(([sectionName, requiredFields]) => {
    const sectionData = analysis[sectionName];

    if (sectionData == null) {
      console.warn(`Missing section: ${sectionName}`);
      return;
    }

    // Специальная обработка gap_analysis
    if (sectionName === 'gap_analysis') {
      if (!Array.isArray(sectionData.gaps)) {
        console.warn(`gap_analysis.gaps should be an array`);
      } else {
        sectionData.gaps.forEach((gap, idx) => {
          // Проверяем, что в каждом gap есть хотя бы поля start_point либо date/price_range
          if (!gap.start_point || !gap.end_point) {
            console.warn(`gap_analysis.gaps[${idx}] missing start_point/end_point`);
          }
          if (!gap.explanation) {
            console.warn(`gap_analysis.gaps[${idx}] missing explanation`);
          }
        });
      }
      return;
    }

    // Массивы объектов
    if (Array.isArray(sectionData)) {
      sectionData.forEach((item, idx) => {
        requiredFields.forEach(field => {
          if (!(field in item)) {
            console.warn(`${sectionName}[${idx}] missing field: ${field}`);
          }
        });
      });
    }
    // Объект с вложенными разделами (pivot_points, indicators_analysis и др.)
    else if (typeof sectionData === 'object') {
      // Если нет конкретных requiredFields — просто проверяем наличие секции
      if (requiredFields.length === 0) {
        return;
      }
      requiredFields.forEach(field => {
        if (!(field in sectionData)) {
          console.warn(`${sectionName} missing field: ${field}`);
        }
      });
    }
    // Иные типы (строки, числа) — проверяем прямо
    else {
      if (requiredFields.length > 0) {
        console.warn(`${sectionName} expected keys ${requiredFields.join(', ')}, but got primitive`);
      }
    }
  });
}
